### PR TITLE
torchcnnbuilder v0.1.0

### PR DESCRIPTION
**Pull request by kdduha**
**Code refactoring**:
- decomposition of functions into different files
- restructuring of namespaces during import
- adding a linter pipeline after each push action in pull-requests
- adding a linter script for the developers
- refactoring all code logic with linter
- updating doc

**New features**:
- new normalization in the convolution block as `InstanceNormNd`
- new adaptive pooling in the decoder as `AdaptiveMaxPoolNd`
- new channels counters like `linear` and `constant`
- `ascending` param is replaced by `channel_growth_rate` which can be `ratio` (same as `ascending = False`), `proportion (same as `ascending = True`), `linear` and `constant`
- new 3d convolutions in `ForecasterBase`

### CODE REVIEW

**Comment by kdduha** [comment]
- [ ] code refactoring is Done

**Comment by kdduha** [comment]
- [ ] Updating ForecasterBase with 3d convolution and refactoring channel counter

**Review by ChrisLisbon — COMMENTED** [review]
В целом заметила, что channel_growth_rate довольно глубоко прокинут, наверное где-то на самом верхнем уровне стоит пояснить поподробнее, в чем между ними различие

File: torchcnnbuilder/utils.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    # by default in all other cases
+    return nn.BatchNorm2d
+
+
+def _select_adaptive_pooling_dimension(conv_dim: int, pooling: str = "avgpool") -> Type[nn.Module]:
+    """
+    The function to select nn.AdaptiveAvgPoolNd
+
+    :param conv_dim: the dimension of the convolutional operation
```
[the remark this answers is not in the document]
Comment (kdduha) [reply torchcnnbuilder/utils.py:107]: Да, это правда, что не поддерживаются. Я везде сделаю ограничение тогда. PoolNd просто в декодер части подгоняет после транспонированных сверток аутпут с последнего слоя под инпут перед энкодером. Там просто тоже разные размерности есть под аналогичные свертки

**Comment by kdduha** [comment]
> В целом заметила, что channel_growth_rate довольно глубоко прокинут, наверное где-то на самом верхнем уровне стоит пояснить поподробнее, в чем между ними различие

Я думаю, это задача документации уже будет, которую я апдейтну - пояснить разницу между разными способами подсчета каналов, в коде не уверен, что такое нужно где-то прописывать повыше

**Comment by ChrisLisbon** [comment]
image
Вот такое еще поймала при дефолтном запуске

**Comment by kdduha** [comment]
> image Вот такое еще поймала при дефолтном запуске

Поленился отдебажить, криво валидацию сделал, спасибо, что нашла!! Пофиксю и сам все попробую ещё раз

**Comment by ChrisLisbon** [comment]
> > image Вот такое еще поймала при дефолтном запуске
> 
> Поленился отдебажить, криво валидацию сделал, спасибо, что нашла!! Пофиксю и сам все попробую ещё раз

https://github.com/ChrisLisbon/TorchCNNBuilder/pull/11/files#diff-3803af975a75051d2742f866d69363d05cbb58ec46bd5b53e97f72ea2cb3073dR72
вот тут заменить uniform и работает

**Comment by kdduha** [comment]
Доку везде апдейтнул, баг(и) поправил + все прогнал еще раз. Вроде все готово, что планировалось
